catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DumpCommand.java

```diff
@@ -407,13 +406,136 @@ protected void printStatus(long count) {
     return metacardTransformerList;
   }
 
-  private Optional<QueryResponseTransformer> getZipCompression() throws InvalidSyntaxException {
-    if (!zipCompression.isPresent()) {
-      zipCompression =
-          getServiceByFilter(
-              QueryResponseTransformer.class,
-              "(|" + "(" + Constants.SERVICE_ID + "=" + ZIP_COMPRESSION + ")" + ")");
+  private void createZip(SourceResponse upstreamResponse, String filePath)
+      throws CatalogTransformerException {
+    if (upstreamResponse.getResults() == null || upstreamResponse.getResults().isEmpty()) {
+      throw new CatalogTransformerException("Source response cannot be null or empty");
+    }
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(filePath);
+        ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
+
+      List<Result> resultList = upstreamResponse.getResults();
+      Map<String, Resource> resourceMap = new HashMap<>();
+
+      // write the metacards to the zip
+      resultList
+          .stream()
+          .map(Result::getMetacard)
+          .forEach(
+              metacard -> {
+                writeMetacardToZip(zipOutputStream, metacard);
+
+                if (hasLocalResources(metacard)) {
+                  resourceMap.putAll(getAllMetacardContent(metacard));
+                }
+              });
+
+      // write the resources to the zip
+      resourceMap.forEach(
+          (filename, resource) -> writeResourceToZip(zipOutputStream, filename, resource));
+
+    } catch (IOException e) {
+      throw new CatalogTransformerException(
+          String.format(
+              "Error occurred when initializing/closing ZipOutputStream with path %s.", filePath),
+          e);
+    }
+  }
+
+  private void writeMetacardToZip(ZipOutputStream zipOutputStream, Metacard metacard) {
+    InputStream inputStream = null;
+
+    try (FileBackedOutputStream fileBackedOutputStream = new FileBackedOutputStream(BUFFER_SIZE);
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileBackedOutputStream)) {
+
+      ZipEntry zipEntry = new ZipEntry(METACARD_PATH + metacard.getId());
+      zipOutputStream.putNextEntry(zipEntry);
+
+      objectOutputStream.writeObject(new MetacardImpl(metacard));
+      inputStream = fileBackedOutputStream.asByteSource().openStream();
+
+      IOUtils.copy(inputStream, zipOutputStream);
+    } catch (IOException e) {
+      LOGGER.debug("Failed to add metacard with id {}.", metacard.getId(), e);
+    } finally {
+      try {
+        inputStream.close();
+      } catch (IOException | NullPointerException ignore) {
+      }
```

Comment:
❓ should we be logging something if we can't close this?

Reply:
Also where is the possible nullpointer? usually we should try to avoid catching a nullpointer and just directly check for null unless the nullpointer is in 3rd party code out of our control or something.

Reply:
Reworked this just a bit and added logging.